File: bindings/erlang/src/proven_math.erl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafeMath - Thin NIF wrapper for overflow-checked arithmetic.
%%
%% This module is a THIN FFI WRAPPER. All computation is performed in
%% formally verified Idris 2 code via the Zig FFI bridge (libproven).
%% No logic is reimplemented in Erlang -- only data marshaling.

-module(proven_math).

-export([
    safe_add/2,
    safe_sub/2,
    safe_mul/2,
    safe_div/2,
    safe_mod/2,
    safe_abs/1,
    clamp/3,
    safe_pow/2
]).

%% @doc Safe addition with overflow checking.
%% Delegates to proven:math_add_checked/2 NIF.
%% Returns {ok, Result} or {error, overflow}.
-spec safe_add(integer(), integer()) -> {ok, integer()} | {error, overflow}.
safe_add(A, B) ->
    proven:math_add_checked(A, B).

%% @doc Safe subtraction with overflow checking.
%% Delegates to proven:math_sub_checked/2 NIF.
%% Returns {ok, Result} or {error, overflow}.
-spec safe_sub(integer(), integer()) -> {ok, integer()} | {error, overflow}.
safe_sub(A, B) ->
    proven:math_sub_checked(A, B).

%% @doc Safe multiplication with overflow checking.
%% Delegates to proven:math_mul_checked/2 NIF.
%% Returns {ok, Result} or {error, overflow}.
-spec safe_mul(integer(), integer()) -> {ok, integer()} | {error, overflow}.
safe_mul(A, B) ->
    proven:math_mul_checked(A, B).

%% @doc Safe division with zero check.
%% Delegates to proven:math_div/2 NIF.
%% Returns {ok, Result} or {error, division_by_zero | overflow}.
-spec safe_div(integer(), integer()) -> {ok, integer()} | {error, division_by_zero | overflow}.
safe_div(A, B) ->
    proven:math_div(A, B).

%% @doc Safe modulo with zero check.
%% Delegates to proven:math_mod/2 NIF.
%% Returns {ok, Result} or {error, division_by_zero}.
-spec safe_mod(integer(), integer()) -> {ok, integer()} | {error, division_by_zero}.
safe_mod(A, B) ->
    proven:math_mod(A, B).

%% @doc Safe absolute value with overflow check.
%% Delegates to proven:math_abs_safe/1 NIF.
%% Returns {ok, Result} or {error, overflow}.
-spec safe_abs(integer()) -> {ok, integer()} | {error, overflow}.
safe_abs(A) ->
    proven:math_abs_safe(A).

%% @doc Clamp a value to a range [Lo, Hi].
%% Delegates to proven:math_clamp/3 NIF.
-spec clamp(number(), number(), number()) -> number().
clamp(Lo, Hi, Value) ->
    proven:math_clamp(Lo, Hi, Value).

%% @doc Safe exponentiation with overflow check.
%% Delegates to proven:math_pow_checked/2 NIF.
%% Returns {ok, Result} or {error, overflow}.
-spec safe_pow(integer(), integer()) -> {ok, integer()} | {error, overflow}.
safe_pow(Base, Exp) ->
    proven:math_pow_checked(Base, Exp).

File: bindings/prolog/proven_math.pl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafeMath - FFI bindings to libproven math operations.
%% All arithmetic is performed in verified Idris 2 code via libproven.

:- module(proven_math, [
    safe_add/3,
    safe_sub/3,
    safe_mul/3,
    safe_div/3,
    safe_mod/3,
    safe_abs/2,
    safe_pow/3,
    clamp/4
]).

:- use_foreign_library(foreign(libproven)).

%% FFI declarations for libproven math functions.
%% Each returns a status code and a value via output arguments.
:- foreign(proven_math_add_checked_ffi, c,
           proven_math_add_checked(+integer, +integer, [-integer], [-integer])).
:- foreign(proven_math_sub_checked_ffi, c,
           proven_math_sub_checked(+integer, +integer, [-integer], [-integer])).
:- foreign(proven_math_mul_checked_ffi, c,
           proven_math_mul_checked(+integer, +integer, [-integer], [-integer])).
:- foreign(proven_math_div_ffi, c,
           proven_math_div(+integer, +integer, [-integer], [-integer])).
:- foreign(proven_math_mod_ffi, c,
           proven_math_mod(+integer, +integer, [-integer], [-integer])).
:- foreign(proven_math_abs_safe_ffi, c,
           proven_math_abs_safe(+integer, [-integer], [-integer])).
:- foreign(proven_math_clamp_ffi, c,
           proven_math_clamp(+integer, +integer, +integer, [-integer])).
:- foreign(proven_math_pow_checked_ffi, c,
           proven_math_pow_checked(+integer, +integer, [-integer], [-integer])).

%! safe_add(+A, +B, -Result) is det.
%
%  Checked addition via libproven.
%  Result is ok(Sum) or error(overflow).
safe_add(A, B, Result) :-
    proven_math_add_checked_ffi(A, B, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(overflow)
    ).

%! safe_sub(+A, +B, -Result) is det.
%
%  Checked subtraction via libproven.
%  Result is ok(Difference) or error(overflow).
safe_sub(A, B, Result) :-
    proven_math_sub_checked_ffi(A, B, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(overflow)
    ).

%! safe_mul(+A, +B, -Result) is det.
%
%  Checked multiplication via libproven.
%  Result is ok(Product) or error(overflow).
safe_mul(A, B, Result) :-
    proven_math_mul_checked_ffi(A, B, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(overflow)
    ).

%! safe_div(+A, +B, -Result) is det.
%
%  Safe division via libproven.
%  Result is ok(Quotient), error(division_by_zero), or error(overflow).
safe_div(A, B, Result) :-
    proven_math_div_ffi(A, B, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Status =:= -5
    ->  Result = error(division_by_zero)
    ;   Result = error(overflow)
    ).

%! safe_mod(+A, +B, -Result) is det.
%
%  Safe modulo via libproven.
%  Result is ok(Remainder) or error(division_by_zero).
safe_mod(A, B, Result) :-
    proven_math_mod_ffi(A, B, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(division_by_zero)
    ).

%! safe_abs(+A, -Result) is det.
%
%  Safe absolute value via libproven.
%  Result is ok(AbsValue) or error(overflow).
safe_abs(A, Result) :-
    proven_math_abs_safe_ffi(A, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(overflow)
    ).

%! safe_pow(+Base, +Exponent, -Result) is det.
%
%  Checked exponentiation via libproven.
%  Result is ok(Power) or error(overflow).
safe_pow(Base, Exp, Result) :-
    proven_math_pow_checked_ffi(Base, Exp, Status, Value),
    (   Status =:= 0
    ->  Result = ok(Value)
    ;   Result = error(overflow)
    ).

%! clamp(+Value, +Lo, +Hi, -Result) is det.
%
%  Clamp value to [Lo, Hi] via libproven.
clamp(Value, Lo, Hi, Result) :-
    proven_math_clamp_ffi(Lo, Hi, Value, Result).
